src/runtime/string.cpp

```diff
@@ -680,6 +680,56 @@ static PRIMFN(prim_hash_str) {
   RETURN(String::alloc(runtime.heap, std::move(hex)));
 }
 
+static PRIMTYPE(type_to_upper) {
+  return args.size() == 1 && args[0]->unify(Data::typeString) && out->unify(Data::typeString);
+}
+
+static PRIMFN(prim_to_upper) {
+  EXPECT(1);
+  STRING(str, 0);
+
+  const char *in = str->c_str();
+  ssize_t in_size = str->length;
+  std::string out;
+  while (in_size > 0) {
+    int32_t codepoint = 0;
+    ssize_t size = utf8proc_iterate(reinterpret_cast<const uint8_t *>(in), in_size, &codepoint);
+    in += size;
+    in_size -= size;
+    int32_t upper = utf8proc_toupper(codepoint);
+    char buf[4] = {0};
```

Comment:
maybe I'm misunderstanding the API, but `4` here doesn't feel safe

Reply:
encode encodes a single code point and returns the number of bytes it took up. The API specifically wants 4 bytes here.